File: frontend/fleet_frontend/src/types/vehicle.ts
export type FuelType = "GASOLINE" | "DIESEL" | "ELECTRIC" | "HYBRID" | "LPG";

export type TransmissionType = "MANUAL" | "AUTOMATIC" | "SEMI_AUTOMATIC";

export type VehicleStatus =
  | "AVAILABLE"
  | "IN_USE"
  | "WITH_DRIVER"
  | "OUTSIDE_DEPOT"
  | "RETURNING_TO_DEPOT"
  | "PARKED"
  | "UNDER_MAINTENANCE"
  | "OUT_OF_SERVICE"
  | "RESERVED"
  | "BROKEN_DOWN";

export interface Vehicle {
  id: number;
  registrationNumber: string;
  brand: string;
  model: string;
  year: number;
  color?: string | null;
  vin?: string | null;
  fuelType?: FuelType | null;
  transmission?: TransmissionType | null;
  status: VehicleStatus;
  parked?: boolean | null;
  mileage?: number | null;
  lastMaintenanceDate?: string | null;
  nextMaintenanceDate?: string | null;
  lastFuelLevel?: number | null;

  currentLatitude?: number | null;
  currentLongitude?: number | null;
  currentCity?: string | null;

  homeDepotCity?: string | null;
  homeDepotLatitude?: number | null;
  homeDepotLongitude?: number | null;

  driverId?: number | null;
  driverName?: string | null;
  driverEmail?: string | null;

  createdAt: string;
  updatedAt: string;
  ownerId?: number | null;
}

export interface VehicleDTO {
  id?: number;
  registrationNumber: string;
  brand: string;
  model: string;
  year: number;
  color?: string | null;
  vin?: string | null;
  fuelType?: FuelType | null;
  transmission?: TransmissionType | null;
  status?: VehicleStatus;
  parked?: boolean | null;
  mileage?: number | null;
  lastMaintenanceDate?: string | null;
  nextMaintenanceDate?: string | null;
  lastFuelLevel?: number | null;

  currentLatitude?: number | null;
  currentLongitude?: number | null;
  currentCity?: string | null;

  homeDepotCity?: string | null;
  homeDepotLatitude?: number | null;
  homeDepotLongitude?: number | null;

  driverId?: number | null;
}

export interface VehicleFilters {
  search?: string;
  status?: VehicleStatus | "ALL";
  fuelType?: FuelType;
  yearFrom?: number;
  yearTo?: number;
  brand?: string;
}

export const VEHICLE_STATUS_OPTIONS = [
  { value: "AVAILABLE", label: "Available" },
  { value: "IN_USE", label: "In use" },
  { value: "RETURNING_TO_DEPOT", label: "Returning to depot" },
  { value: "UNDER_MAINTENANCE", label: "Under maintenance" },
  { value: "BROKEN_DOWN", label: "Broken down" },
  { value: "OUT_OF_SERVICE", label: "Out of service" },
  { value: "RESERVED", label: "Reserved" },
] as const;

export function getVehicleBusinessLabel(vehicle: Pick<Vehicle, "status" | "parked">) {
  if (vehicle.status === "AVAILABLE" && vehicle.parked) return "Available at depot";
  if (vehicle.status === "AVAILABLE" && !vehicle.parked) return "Available on field";
  if (vehicle.status === "RETURNING_TO_DEPOT") return "Returning to depot";
  if (vehicle.status === "IN_USE") return "In mission";
  if (vehicle.status === "BROKEN_DOWN") return "Broken down";
  if (vehicle.status === "UNDER_MAINTENANCE") return "Under maintenance";
  if (vehicle.status === "OUT_OF_SERVICE") return "Out of service";
  return vehicle.status;
}

export const getVehicleStatusColor = (status: VehicleStatus): string => {
  const colors: Record<VehicleStatus, string> = {
    AVAILABLE: "bg-green-100 text-green-800 border-green-200",
    IN_USE: "bg-blue-100 text-blue-800 border-blue-200",
    WITH_DRIVER: "bg-cyan-100 text-cyan-800 border-cyan-200",
    OUTSIDE_DEPOT: "bg-orange-100 text-orange-800 border-orange-200",
    RETURNING_TO_DEPOT: "bg-purple-100 text-purple-800 border-purple-200",
    PARKED: "bg-emerald-100 text-emerald-800 border-emerald-200",
    UNDER_MAINTENANCE: "bg-yellow-100 text-yellow-800 border-yellow-200",
    OUT_OF_SERVICE: "bg-red-100 text-red-800 border-red-200",
    RESERVED: "bg-slate-100 text-slate-800 border-slate-200",
    BROKEN_DOWN: "bg-red-100 text-red-800 border-red-200",
  };

  return colors[status] || "bg-gray-100 text-gray-800 border-gray-200";
};

File: fleet_backend/src/main/java/com/example/fleet_backend/model/Vehicle.java
package com.example.fleet_backend.model;

import jakarta.persistence.*;
import java.time.LocalDateTime;

/**
 * ✅ Entité Vehicle
 *
 * Représente un véhicule dans le système de gestion de flotte.
 *
 * Cette classe est mappée à la table "vehicles" en base de données.
 *
 * Relations principales :
 * - ManyToOne vers Driver (conducteur assigné)
 * - ManyToOne vers User (owner/propriétaire du véhicule)
 *
 * Contient :
 * - Informations d’identification (registrationNumber, vin)
 * - Caractéristiques techniques (fuelType, transmission)
 * - État métier (status, mileage)
 * - Données maintenance
 * - Audit (createdAt, updatedAt)
 */
@Entity
@Table(name = "vehicles")
public class Vehicle {

    /**
     * ✅ Clé primaire auto-générée
     */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * ✅ Numéro d'immatriculation
     * - Obligatoire
     * - Unique
     */
    @Column(name = "registration_number", nullable = false, unique = true)
    private String registrationNumber;

    /**
     * ✅ Marque du véhicule (ex: Toyota, BMW)
     */
    @Column(name = "brand", nullable = false)
    private String brand;

    /**
     * ✅ Modèle du véhicule
     */
    @Column(name = "model", nullable = false)
    private String model;

    /**
     * ✅ Année de fabrication
     */
    @Column(name = "year", nullable = false)
    private Integer year;

    /**
     * ✅ Couleur (optionnel)
     */
    @Column(name = "color")
    private String color;

    /**
     * ✅ VIN (Vehicle Identification Number)
     * - Unique si présent
     */
    @Column(name = "vin", unique = true)
    private String vin;

    /**
     * ✅ Type de carburant
     * Enum stocké en STRING en base
     */
    @Column(name = "fuel_type")
    @Enumerated(EnumType.STRING)
    private FuelType fuelType;

    /**
     * ✅ Type de transmission
     */
    @Column(name = "transmission")
    @Enumerated(EnumType.STRING)
    private TransmissionType transmission;

    /**
     * ✅ Statut métier du véhicule
     * Exemple : AVAILABLE, IN_USE, UNDER_MAINTENANCE
     */
    @Column(name = "status")
    @Enumerated(EnumType.STRING)
    private VehicleStatus status;

    /**
     * ✅ Kilométrage actuel
     */
    @Column(name = "mileage")
    private Double mileage;

    /**
     * ✅ Date dernière maintenance
     */
    @Column(name = "last_maintenance_date")
    private LocalDateTime lastMaintenanceDate;

    /**
     * ✅ Date prochaine maintenance prévue
     */
    @Column(name = "next_maintenance_date")
    private LocalDateTime nextMaintenanceDate;

    /**
     * ✅ Conducteur assigné
     *
     * ManyToOne :
     * - Plusieurs véhicules peuvent être assignés à un même driver
     *
     * FetchType.EAGER :
     * - Le driver est chargé automatiquement avec le véhicule
     */
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "driver_id")
    private Driver driver;

    /**
     * ✅ Owner (propriétaire du véhicule)
     *
     * ManyToOne :
     * - Un owner peut posséder plusieurs véhicules
     *
     * FetchType.LAZY :
     * - L'owner est chargé seulement si nécessaire
     *
     * nullable = false :
     * - Chaque véhicule doit obligatoirement appartenir à un owner
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id", nullable = false)
    private User owner;

    /**
     * ✅ Date création (audit)
     */
    @Column(name = "created_at")
    private LocalDateTime createdAt;

    /**
     * ✅ Date dernière mise à jour (audit)
     */
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    /**
     * Constructeur vide requis par JPA
     */
    public Vehicle() {}

    /**
     * Constructeur pratique pour création rapide
     */
    public Vehicle(String registrationNumber, String brand, String model, Integer year) {
        this.registrationNumber = registrationNumber;
        this.brand = brand;
        this.model = model;
        this.year = year;
    }

    /**
     * ✅ Méthode appelée automatiquement avant INSERT en base.
     *
     * - Initialise createdAt
     * - Initialise updatedAt
     * - Définit status par défaut = AVAILABLE
     * - Définit mileage par défaut = 0.0
     */
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();

        if (status == null) {
            status = VehicleStatus.AVAILABLE;
        }
        if (mileage == null) {
            mileage = 0.0;
        }
    }

    /**
     * ✅ Méthode appelée automatiquement avant UPDATE.
     *
     * Met à jour updatedAt.
     */
    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    // =========================
    // GETTERS & SETTERS
    // =========================

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public String getRegistrationNumber() { return registrationNumber; }
    public void setRegistrationNumber(String registrationNumber) { this.registrationNumber = registrationNumber; }

    public String getBrand() { return brand; }
    public void setBrand(String brand) { this.brand = brand; }

    public String getModel() { return model; }
    public void setModel(String model) { this.model = model; }

    public Integer getYear() { return year; }
    public void setYear(Integer year) { this.year = year; }

    public String getColor() { return color; }
    public void setColor(String color) { this.color = color; }

    public String getVin() { return vin; }
    public void setVin(String vin) { this.vin = vin; }

    public FuelType getFuelType() { return fuelType; }
    public void setFuelType(FuelType fuelType) { this.fuelType = fuelType; }

    public TransmissionType getTransmission() { return transmission; }
    public void setTransmission(TransmissionType transmission) { this.transmission = transmission; }

    public VehicleStatus getStatus() { return status; }
    public void setStatus(VehicleStatus status) { this.status = status; }

    public Double getMileage() { return mileage; }
    public void setMileage(Double mileage) { this.mileage = mileage; }

    public LocalDateTime getLastMaintenanceDate() { return lastMaintenanceDate; }
    public void setLastMaintenanceDate(LocalDateTime lastMaintenanceDate) { this.lastMaintenanceDate = lastMaintenanceDate; }

    public LocalDateTime getNextMaintenanceDate() { return nextMaintenanceDate; }
    public void setNextMaintenanceDate(LocalDateTime nextMaintenanceDate) { this.nextMaintenanceDate = nextMaintenanceDate; }

    public Driver getDriver() { return driver; }
    public void setDriver(Driver driver) { this.driver = driver; }

    public User getOwner() { return owner; }
    public void setOwner(User owner) { this.owner = owner; }

    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }

    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }

    // =========================
    // ENUMS
    // =========================

    /**
     * Types de carburant supportés.
     */
    public enum FuelType {
        GASOLINE, DIESEL, ELECTRIC, HYBRID, LPG
    }

    /**
     * Types de transmission.
     */
    public enum TransmissionType {
        MANUAL, AUTOMATIC, SEMI_AUTOMATIC
    }

    /**
     * Statuts métier du véhicule.
     */
    public enum VehicleStatus {
        AVAILABLE,
        IN_USE,
        UNDER_MAINTENANCE,
        OUT_OF_SERVICE,
        RESERVED
    }
}